docs/src/main/paradox/master-in-detail.md

```diff
@@ -0,0 +1,50 @@
+
+This allows us to capture and store each such event that happens, we can later replaying all of them on an empty model and arrive at the exact same state. This is how event sourcing and [Akka Persistence](http://doc.akka.io/docs/akka/current/scala/persistence.html) allows us to start the actor (possibly on a different node) and reach the same state as a previous instance.
+
+Inside of the actor this means that any time the `WorkState` is modified, we must first `persist` the event before acting on it. Not until we know the event was successfully stored we can apply the modification to the state. If we would apply it before persisting we could apply a change, have the write fail, and be in a state a replaying node could not reach.
+
+Let's look at how a command to process a work item from the front-end comes in:
+
+@@snip [Master.scala]($g8src$/scala/worker/Master.scala) { #persisting }
+
+The first thing you might notice is the comment saying _idempotent_, this means that the same work message may arrive multiple times, but regardless how many times the same work arrives, it should only be executed once. This is needed since the `FrontEnd` actor re-sends work in case of the `Work` or `Ack` messages getting lost (Akka does not provide any guarantee of delivery, [see details in the docs](http://doc.akka.io/docs/akka/current/scala/general/message-delivery-reliability.html#discussion-why-no-guaranteed-delivery-)).
+
+To make the logic idempotent we simple check if the work id is already known, and if it is we simply `Ack` it without further logic. If the work is previously unknown, we start by transforming it into a `WorkAccepted` event, which we persist, and only in the `handler`-function passed to `persist` do we actually update the `workState`, send an `Ack` back to the `FrontEnd` and trigger a search for available workers.
+
+
+## Akka Persistence differences from regular Actors
+
+While in a "normal" Actor the only thing we have to do is to implement `receive`, which is then invoked for each incoming message. In a `PersistentActor` there are three things that needs to be implemented:
+
+ 1. `persistenceId` is a global identifier for the actor, we must make sure that there is never more than one Actor instance with the same `persistenceId` running globally, or else we would possibly mess up its journal.
+ 1. `receiveCommand` corresponds to the `receive` method of regular actors. Messages sent to the actor end up here. 
+ 1. `receiveRecover` is invoked with the recorded events of the actor when it starts up 
+
+## Keeping track of workers
+
+For the set of worker, saving them to a persistent store would be less valuable, as there is a high chance that the set of workers and worker nodes are not the same after a stop and start of the master. Imagine that we stopped an entire cluster and started a new one, then the entire set of workers would be invalid and the cluster would have to start by figuring this out and invalidating the workers. 
```

Comment:
Maybe instead:
Unlike the 'Master' actor, the example system contains multiple workers that can be stopped and restarted frequently. We do not need to save their state since the 'Master' is tracking work and will simply send work to another worker if the original fails to respond. So, rather than persisting a list of available workers, the example uses the following strategy: 

- Running workers periodically register with the master using a 'RegisterWorker' message. If a 'back-end' node fails and the 'Master' is started on a new node, the registrations go automatically to the new node.
- Any type of failure -- whether from the network, worker actor, or node -- that prevents a 'RegisterWorker' message from arriving within the 'work-timeout' period causes the 'Master' actor to remove the worker from its list.